crates/bevy_math/src/curve/mod.rs

```diff
@@ -0,0 +1,610 @@
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.reparam_curve.domain()
+    }
+
+    #[inline]
+    fn sample(&self, t: f32) -> T {
+        let sample_time = self.reparam_curve.sample(t);
+        self.base.sample(sample_time)
+    }
+}
+
+/// A curve that is the graph of another curve over its parameter space.
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
+pub struct GraphCurve<T, C> {
+    base: C,
+    _phantom: PhantomData<T>,
+}
+
+impl<T, C> Curve<(f32, T)> for GraphCurve<T, C>
+where
+    C: Curve<T>,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.base.domain()
+    }
+
+    #[inline]
+    fn sample(&self, t: f32) -> (f32, T) {
+        (t, self.base.sample(t))
+    }
+}
+
+/// A curve that combines the data from two constituent curves into a tuple output type.
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
+pub struct ProductCurve<S, T, C, D> {
+    domain: Interval,
+    first: C,
+    second: D,
+    _phantom: PhantomData<(S, T)>,
+}
+
+impl<S, T, C, D> Curve<(S, T)> for ProductCurve<S, T, C, D>
+where
+    C: Curve<S>,
+    D: Curve<T>,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.domain
+    }
+
+    #[inline]
+    fn sample(&self, t: f32) -> (S, T) {
+        (self.first.sample(t), self.second.sample(t))
+    }
+}
+
+/// The curve that results from chaining one curve with another. The second curve is
+/// effectively reparametrized so that its start is at the end of the first.
+///
+/// For this to be well-formed, the first curve's domain must be right-finite and the second's
+/// must be left-finite.
+pub struct ChainCurve<T, C, D> {
+    first: C,
+    second: D,
+    _phantom: PhantomData<T>,
+}
+
+impl<T, C, D> Curve<T> for ChainCurve<T, C, D>
+where
+    C: Curve<T>,
+    D: Curve<T>,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        // This unwrap always succeeds because `first` has a valid Interval as its domain and the
+        // length of `second` cannot be NAN. It's still fine if it's infinity.
+        Interval::new(
+            self.first.domain().start(),
+            self.first.domain().end() + self.second.domain().length(),
+        )
+        .unwrap()
+    }
+
+    #[inline]
+    fn sample(&self, t: f32) -> T {
+        if t > self.first.domain().end() {
+            self.second.sample(
+                // `t - first.domain.end` computes the offset into the domain of the second.
+                t - self.first.domain().end() + self.second.domain().start(),
+            )
+        } else {
+            self.first.sample(t)
+        }
+    }
+}
+
+/// Create a [`Curve`] that constantly takes the given `value` over the given `domain`.
+pub fn constant_curve<T: Clone>(domain: Interval, value: T) -> ConstantCurve<T> {
```

Comment:
if anything i would expect the canonical way to construct this to be `Curve::constant(value)`. i'm not sure if that's possible because of object safety, but if not, i would simply rename this function `constant`